SAK-29825 New service to set the gradebook grading scales.

ddelblanco opened the pull request:
SAK-29825 New service to set the gradebook grading scales. 

This time without the SQLService, only using the edu-services/gradebook services. 

Sample call:

http://SERVERNAME/sakai-ws/rest/gradebook/createOrUpdateGradeScale?sessionid=THESESSIONID&scaleUuid=27&scaleName=newscale&grades={eee,bbb,ccc}&percents={69,31,12}&updateOld=true&updateOnlyNotCustomized=true

updateOld=true updates the old values in the actual gradebooks. It depens of updateOnlyNotCustomized=true if it updates all of them or only the ones that has never been customized by instructors.
updateOld false only update the scale definition, but not the actual gradebooks.

steveswinsburg commented on edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookFrameworkService.java:
```diff
@@ -79,4 +80,31 @@
+	 *   The gradeMApping to update.
+	 *
+	 */
+	public void updateGradeMapping(Object gradeMapping);
```
Could these types be a bit more explicit?

ddelblanco replied:
Yes, I agree that passing the object is not a great thing, but GradeMapping is not defined in the API. The API is a dependence, so if I try to import these things from the hibernate implementation, it, logically, generates a cyclic reference. The only way I've found is to pass as object and later in the implementation convert it to the right type.   

It is done in this other method in the same place:
/**
     * @param gradingScaleDefinitions
     *	A collection of GradingScaleDefinition beans.
	 */
	public void setAvailableGradingScales(Collection gradingScaleDefinitions);

It passed a Collection of gradinScaleDefinitions but they are objects... not really gradinScaleDefinitions... because this is not in the API.

I'm open to any suggestion to change this but of course, add this to the API is a big work, becuase GradeMappindg has a Gradebook and a GradingScale "inside" so I would need to add that to the API too... and surely other things after that in cascade...

steveswinsburg replied:
I would be following the other patterns of creating a DTO, eg shared.Assignment, shared.CategoryDefinition etc for passing data around, then mapping it between the DTO and hibernate objects.

ddelblanco replied:
Finally I've decided to pass as parameters the GradeMapping ID, the MAP with the new grades and update it, in this way, the API don't need to be changed.

steveswinsburg commented on edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java:
```diff
@@ -743,4 +743,12 @@ public void finalizeGrades(String gradebookUid)
+	 * @param id
+	 * @return a gradebook with all the GradeMappings.
+	 */
+	public Object getGradebookWithGradeMappings(Long id);
```
Can you return just the GradeMapping?

ddelblanco replied:
This is the same problem that I have in the comment before.

steveswinsburg replied:
OK so I would be making a shared DTO and mapping these hibernate objects to that and back again.

steveswinsburg commented on edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookFrameworkService.java:
```diff
@@ -79,4 +80,31 @@
+	 *   The gradebook with GradeMappings where we will add the grading scale.
+	 *
+	 */
+	public void saveGradeMappingToGradebook(String scaleUuid, Object gradebook);
```
Most of the GB calls pass in the Gradebook uid, not the entire object. Could this be made to be the same? The code would then look it up when it is run. External tools won't have the Gradebook, just the uid reference to it.

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java:
```diff
@@ -188,6 +188,60 @@ public Object doInHibernate(Session session) throws HibernateException {
+
+	public List getAvailableGradingScales() {
+
+		List gradingScales2 = (List)getHibernateTemplate().execute(new HibernateCallback() {
```
Could you call this 'rval' or something? It's not immediately obvious why its '2'.

ddelblanco replied:
That's easy!!! :)

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:
```diff
@@ -2815,6 +2815,17 @@ public void setGradebookPermissionService(GradebookPermissionService gradebookPe
 
-	
+	@Override
+	public Object getGradebookWithGradeMappings(final Long id) {
```
As above, I think it would be more useful to return the GradeMappings instead of the entire Gradebook.

ddelblanco replied:
Finally I've changed the logic and avoid some of the iterators. 

Now, first I return the sitesIds and not the sites, surely saving some memory.
Later, I don't create a gradebook list, I directly work on the gradebook in the siteid's iterator if the site has gradebook.
if it is a new scale, passing scale Uuid and gradebook Uid I update all.
if it is a update, then I need to pass the id to recuperate the gradeMappings with this method: getGradebookGradeMappings(final Long gradebookId)  I need to pass the id and not the Uid to make it work. Compare gradings and update if needed and based on the booleans.

I think now is more efficient and fast. 

As now is working fine, I'm going to update the pull request with this changes, and then start to work in the API thing.

ddelblanco commented:
I've solved all the comments, except the new API classes... I'll work on this from now.

ddelblanco commented:
I've solved the API problem too passing other parameters. So now all should be solved.

steveswinsburg commented on webservices/cxf/src/java/org/sakaiproject/webservices/SakaiGradebook.java:
```diff
@@ -0,0 +1,235 @@
+        return "success";
+    }
+
+    private GradingScaleDefinition toGradingScaleDefinition(GradingScale gScale){
```
This should be in the API so others can use it.

steveswinsburg commented on webservices/cxf/src/java/org/sakaiproject/webservices/SakaiGradebook.java:
```diff
@@ -0,0 +1,235 @@
+
+            //Compare the UID of the scale to check if we need to update ot create a new one
+            for (Iterator iter = gradingScales.iterator(); iter.hasNext();) {
+                GradingScaleDefinition gradingScale = toGradingScaleDefinition((GradingScale)iter.next());
```
Could you move this mapping into the service?

steveswinsburg commented on webservices/cxf/src/java/org/sakaiproject/webservices/SakaiGradebook.java:
```diff
@@ -0,0 +1,235 @@
+        return scaleDef;
+    }
+
+    private GradingScale toGradingScale(GradingScaleDefinition bean) {
```
This mapping also.

steveswinsburg commented:
This is good but a few things need to be moved to the GB service and out of the web service.

ddelblanco commented:
I've added the conversion from gradingScale "toGradingScaleDefinition" to the GradingScale object so it can be used by anyone, and I've removed the unused toGradingScale from GradingScaleDefinition becuase was not needed.

steveswinsburg commented:
Hey Diego, there are some conflicts on this branch. Can you fix and then I can review and merge?

ddelblanco commented:
rebased!

steveswinsburg commented on edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookService.java:
```diff
@@ -781,4 +782,11 @@ public void finalizeGrades(String gradebookUid)
+	 * @param gradebookId
+	 * @return a gradebook with all the GradeMappings.
+	 */
+	public Set getGradebookGradeMappings(Long gradebookId);
```
It would be nice if this also took a gradebookUid, but I'll add that.

steveswinsburg replied:
Also, javadocs out of date here. I'll correct them.

steveswinsburg replied:
Ideally this shouldn't pass around the persistent objects either, but there isn't a clear separation of this in GradebookService and this one is a bit more complex to map out. Will leave this for now.

steveswinsburg commented on edu-services/gradebook-service/hibernate/src/java/org/sakaiproject/tool/gradebook/GradingScale.java:
```diff
@@ -97,4 +100,25 @@ public String toString() {
             append(getUid()).toString();
     }
+
+	public GradingScaleDefinition toGradingScaleDefinition(){
```
need javadoc

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookFrameworkServiceImpl.java:
```diff
@@ -191,6 +191,61 @@ public Object doInHibernate(Session session) throws HibernateException {
 		});
 	}
 
+	public void updateGradeMapping(Long gradeMappingId, Map<String, Double> gradeMap){
```
Missing @Override annotations on these three methods.

steveswinsburg commented on edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookFrameworkService.java:
```diff
@@ -79,4 +81,31 @@
+	 *   The gradeMApping to update.
+	 *
+	 */
+	public void updateGradeMapping(Long gradeMappingId, Map<String, Double> gradeMap);
```
javadocs need updating, params no longer match. I'll update.

steveswinsburg commented:
Looks good. There are a still a bunch of things to be done for completeness, however I'll be sending a PR for a similar area soon and will do them as part of that.
